node/node.go

```diff
@@ -132,19 +132,23 @@ func NewNode(config *cfg.Config,
 	if err != nil {
 		return nil, err
 	}
+
+	genDoc, err := genesisDocProvider()
```

Comment:
this assumes the genesis will always be available, even if we've already run the node. it's probably ok, but not sure we want to make that assumption

Reply:
@jaekwon do you want to comment on this?

Reply:
If we start a node, and then later the genesis file is changed, we will not be able to restart the node unless we can recover that genesis file. So I think it's worth having the genesis saved in the database to prevent a certain class of user errors. What do you think?

Reply:
don't know. I am trying to come up with a scenario where it will be useful.

## 1. user loses state
then we'll need original genesis file to MakeGenesisState, but since we store all DBs together, a user will most likely lose genesis state too, so no point of storing it

## 2. user accidentally changes the genesis json (or some hacker changes it)
then TM will either fail to start due to a parsing error / missing genesis file error OR RPC will be serving incorrect genesis file. Is it worth saving it in the DB for the purpose of preventing such errors?

Reply:
```
Date:   Fri Oct 13 13:11:16 2017 +0400

    save genesis doc in DB to prevent user errors

    https://github.com/tendermint/tendermint/pull/676#discussion_r144411458

─────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────
modified: node/node.go
─────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────
@ node.go:5 @ package node

import (
    "bytes"
    "encoding/json"
    "errors"
    "fmt"
    "net"
@ node.go:137 @ func NewNode(config *cfg.Config,
        return nil, err
    }

    genDoc, err := genesisDocProvider()
    // Get genesis doc
    genesisDB, err := dbProvider(&DBContext{"genesis", config})
    if err != nil {
        return nil, err
    }
    genDoc, err := loadGenesisDoc(genesisDB)
    if err != nil {
        genDoc, err = genesisDocProvider()
        if err != nil {
            return nil, err
        }
        // save genesis doc to prevent a certain class of user errors (e.g. when it
        // was changed, accidentally or not)
        err = saveGenesisDoc(genesisDB, genDoc)
        if err != nil {
            return nil, err
        }
    }

    state := sm.LoadState(stateDB)
    if state == nil {
@ node.go:554 @ func (n *Node) DialSeeds(seeds []string) error {
}

//------------------------------------------------------------------------------

var (
    genesisDocKey = []byte("genesisDoc")
)

func loadGenesisDoc(db dbm.DB) (*types.GenesisDoc, error) {
    bytes := db.Get(genesisDocKey)
    if len(bytes) == 0 {
        return nil, errors.New("Genesis doc not found")
    } else {
        var genDoc *types.GenesisDoc
        err := json.Unmarshal(bytes, &genDoc)
        return genDoc, err
    }
}

func saveGenesisDoc(db dbm.DB, genDoc *types.GenesisDoc) error {
    bytes, err := json.Marshal(genDoc)
    if err != nil {
        return err
    }
    db.SetSync(genesisDocKey, bytes)
    return nil
}
```

I wrote the code and can push it if we agree that we need to save genesis doc.

Reply:
@silasdavis I feel like you have opinions here

Reply:
I'm not actually sure about passing `GenesisDoc` via a factory method (i.e. `GenesisDocProvider`). Almost by definition `GenesisDoc` creation is point-like and immutable, but a factory method implies it that it may need to created multiple times, lazily initialised, or icreated n different contexts that branch. None of these should be true for genesis. I think it would be better to make `NewNode` take a plain `GenesisDoc` as a argument (value type to emphasise the immutability.) Callers will just need to call your default provider rather than passing it. All I do for the provider is the busywork of wrapping my `GenesisDoc` in a func.

I also think that the `GenesisDoc` is an interesting part of the audit trail of the chain and you should probbly store it and serve it from state reliably.

Also in the world where we are handing provision of genesis doc and database provider to the caller we shouldn't make assumptions about what state is co-located or not or even stored to disk.

Come to think of it I actually think a hash of the `GenesisDoc` should be mixed into the initial `AppHash`. I don't think it's the same chain if it didn't start with the same `GenesisDoc` and validator set, but I suppose the way you have it now allows for some flexibility for the ABCI app in that respect (in next Burrow the initial AppHash is our GenesisDoc hash which includes validator set and time, ChainID is also derived from that)

Reply:
wow, thanks for your input! ok, let's persist genesis doc

Reply:
As for `NewNode` taking `GenesisDoc` as an argument, don't think this is what we want in the end. We call `genesisDocProvider` only once, then we save genesis doc to the state. If we rewrite the code to allow to pass it as an argument, this would mean the caller would have to retrieve it from some source on every start and that's not what we want.

Reply:
what if we change the DBProvider to an interface:

```
type DBProvider interface{
  DB(*DBContext) (dbm.DB, error)
  GenesisDoc() GenesisDoc
}
```

Reply:
meh, this is kinda weird.

Ok I think what we currently have is fine. Let's let it lie for a bit until something genuinely better emerges.